src/schrodinger/rdkit_extensions/helm/monomer_coordgen.cpp

```diff
@@ -758,6 +758,184 @@ void lay_out_polymers(const std::vector<RDKit::ROMOL_SPTR>& polymers)
+    };
+
+    int o1 = orientation(p1, p2, q1);
+    int o2 = orientation(p1, p2, q2);
+    int o3 = orientation(q1, q2, p1);
+    int o4 = orientation(q1, q2, p2);
+
+    // General case
+    if (o1 != o2 && o3 != o4)
+        return true;
+
+    // Special collinear cases
+    if (o1 == 0 && on_segment(p1, q1, p2))
+        return true;
+    if (o2 == 0 && on_segment(p1, q2, p2))
+        return true;
+    if (o3 == 0 && on_segment(q1, p1, q2))
+        return true;
+    if (o4 == 0 && on_segment(q1, p2, q2))
+        return true;
+
+    return false;
+}
+
+static double distance_segments(const RDGeom::Point3D& p1,
+                                const RDGeom::Point3D& p2,
+                                const RDGeom::Point3D& q1,
+                                const RDGeom::Point3D& q2)
+{
+    auto cross2d = [](double x1, double y1, double x2, double y2) {
+        return x1 * y2 - y1 * x2;
+    };
+
+    auto dist_point_to_segment = [](const RDGeom::Point3D& p,
+                                    const RDGeom::Point3D& a,
+                                    const RDGeom::Point3D& b) {
+        double abx = b.x - a.x;
+        double aby = b.y - a.y;
+        double apx = p.x - a.x;
+        double apy = p.y - a.y;
+        double ab_len2 = abx * abx + aby * aby;
+        if (ab_len2 < 1e-18) {
+            double dx = p.x - a.x;
+            double dy = p.y - a.y;
+            return std::sqrt(dx * dx + dy * dy);
+        }
+        double t = (apx * abx + apy * aby) / ab_len2;
+        t = std::clamp(t, 0.0, 1.0);
+        double projx = a.x + t * abx;
+        double projy = a.y + t * aby;
+        double dx = p.x - projx;
+        double dy = p.y - projy;
+        return std::sqrt(dx * dx + dy * dy);
+    };
+
+    // Check 2D segment intersection
+    double r_px = p2.x - p1.x;
+    double r_py = p2.y - p1.y;
+    double s_px = q2.x - q1.x;
+    double s_py = q2.y - q1.y;
+
+    double denom = cross2d(r_px, r_py, s_px, s_py);
+    double qp_x = q1.x - p1.x;
+    double qp_y = q1.y - p1.y;
+
+    if (std::fabs(denom) > 1e-18) {
+        double t = cross2d(qp_x, qp_y, s_px, s_py) / denom;
+        double u = cross2d(qp_x, qp_y, r_px, r_py) / denom;
+        if (t >= 0 && t <= 1 && u >= 0 && u <= 1) {
+            // Segments intersect
+            return 0.0;
+        }
+    }
+
+    // Otherwise, no intersection → compute closest approach
+    double d1 = dist_point_to_segment(p1, q1, q2);
+    double d2 = dist_point_to_segment(p2, q1, q2);
+    double d3 = dist_point_to_segment(q1, p1, p2);
+    double d4 = dist_point_to_segment(q2, p1, p2);
+    return std::min({d1, d2, d3, d4});
+}
+
+static bool check_bond_crossing(const RDKit::ROMol& monomer_mol)
+{
+    const double CLASH_DISTANCE = MONOMER_BOND_LENGTH * 0.25;
+    auto& conformer = monomer_mol.getConformer();
+    auto positions = conformer.getPositions();
+    for (size_t i = 0; i < monomer_mol.getNumBonds(); i++) {
+        auto bond1 = monomer_mol.getBondWithIdx(i);
```

Comment:
This could be `for (auto bond1 : monomer_mol.bonds())`.  Same with the `bond2` loop below.

Reply:
same reason as before.. I think I'd prefer code to be a bit more performant if we can avoid obvious checks